client/src/boot/BootRoutes.js

```diff
@@ -140,8 +158,79 @@ class BootRoutes {
       });
     });
 
+    // Legacy support – make sure we don't overwrite the `onbeforeunload`
+    // handler in `LeftAndMain.EditForm.js` but run it after the new handler.
+    const currBeforeUnload = window.onbeforeunload;
+    window.onbeforeunload = () => {
+      if (this.shouldConfirmBeforeUnload()) {
+        return i18n._t('Admin.CONFIRMUNSAVEDSHORT', 'WARNING: Your changes have not been saved.');
+      }
+
+      if (typeof currBeforeUnload === 'function') {
+        return currBeforeUnload();
+      }
+
+      return null;
+    };
+
     pageRouter.start();
   }
+
+  /**
+   * Return `true` if there are forms with unsaved changes. Return `false`
+   * otherwise.
+   *
+   * @return {Boolean}
+   */
+  shouldConfirmBeforeUnload() {
+    const state = this.store.getState();
+    const unsaveds = state.unsavedForms || {};
```

Comment:
An oversight here is that not every unsaved form should block navigation if un-submitted; E.g. search forms.